FBRetainCycleDetector/Layout/Classes/Parser/FBStructEncodingParser.mm

```diff
@@ -9,95 +9,152 @@
-        nameFromBefore = [types lastObject];
-        [types removeLastObject];
+      index += stringToScan.length();
+      return true;
+    }
+    
+    std::string scanUpToString(std::string upToString) {
+      size_t pos = string.find(upToString, index);
+      if (pos == std::string::npos) {
+        // Mark as whole string scanned
+        index = string.length();
+        return "";
       }
-      FBParsedType *type = [[FBParsedType alloc] initWithName:nameFromBefore
-                                                 typeEncoding:parseResult];
-      [types addObject:type];
+      
+      std::string inBetweenString = string.substr(index, pos - index);
+      index = pos;
+      return inBetweenString;
     }
-  }
-
-  NSPredicate *filterPredicate = [NSPredicate predicateWithBlock:^BOOL(id evaluatedObject,
-                                                                       NSDictionary<NSString *, id> *bindings) {
-    if ([evaluatedObject isKindOfClass:[FBParsedType class]] ||
-        [evaluatedObject isKindOfClass:[FBParsedStruct class]]) {
-      return YES;
+    
+    char currentCharacter() {
+      return string[index];
+    }
+    
+    std::string scanUpToCharacterFromSet(std::string &characterSet) {
+      size_t pos = string.find_first_of(characterSet, index);
+      if (pos == std::string::npos) {
+        index = string.length();
+        return "";
+      }
+      
+      std::string inBetweenString = string.substr(index, pos-index);
+      index = pos;
+      return inBetweenString;
     }
-    return NO;
-  }];
-  NSArray<FBParsedType *> *finalTypes = [types filteredArrayUsingPredicate:filterPredicate];
-
-  return {
-    .containedTypes = finalTypes,
-    .typeName = structTypeName,
   };
-}
-
-FBParsedStruct *FBParseStructEncoding(NSString *structEncodingString) {
-  return FBParseStructEncodingWithName(structEncodingString, nil);
-}
+  
+};
 
-FBParsedStruct *FBParseStructEncodingWithName(NSString *structEncodingString, NSString *structName) {
-  struct _StructParseResult result = _ParseStructEncodingWithScanner([NSScanner scannerWithString:structEncodingString]);
-  FBParsedStruct *outerStruct = [[FBParsedStruct alloc] initWithName:structName
-                                                        typeEncoding:structEncodingString
-                                                      structTypeName:result.typeName
-                                              typesContainedInStruct:result.containedTypes];
-  [outerStruct passTypePath:nil];
-  return outerStruct;
-}
+namespace FB { namespace RetainCycleDetector { namespace Parser {
+  
+  /**
+   Intermediate struct object used inside the algorithm to pass some
+   information when parsing nested structures.
+   */
+  struct _StructParseResult {
+    std::vector<std::shared_ptr<Type>> containedTypes;
+    std::string typeName;
+  };
+  
+  static struct _StructParseResult _ParseStructEncodingWithScanner(_StringScanner &scanner) {
```

Comment:
In C++, the `struct` keyword isn't necessary when referring to `struct` types, like it is in C (and ObjC).